### PR TITLE
Second Part of Internal API Refactor

### PR DESCRIPTION
**Pull request by pcmoritz**
This PR switches the old object IDs with the new ones as a preparation for integrating the GCS APIs.

### CODE REVIEW

File: src/common/state/error_table.cc [hunk 1]
```diff
@@ -22,10 +22,10 @@ void push_error(DBHandle *db_handle,
   info->data_length = data_length;
   memcpy(info->data, data, data_length);
   /* Generate a random key to identify this error message. */
-  CHECK(sizeof(info->error_key) >= UNIQUE_ID_SIZE);
-  UniqueID error_key = globally_unique_id();
-  memcpy(info->error_key, error_key.id, sizeof(info->error_key));
```
Comment (pcmoritz) [inline src/common/state/error_table.cc]: TODO: This is a little strange, investigate!

File: src/global_scheduler/global_scheduler.cc [hunk 2]
```diff
@@ -296,10 +293,8 @@ void object_table_subscribe_callback(ObjectID object_id,
                                      void *user_context) {
   /* Extract global scheduler state from the callback context. */
   GlobalSchedulerState *state = (GlobalSchedulerState *) user_context;
-  char id_string[ID_STRING_SIZE];
-  LOG_DEBUG("object table subscribe callback for OBJECT = %s",
-            ObjectID_to_string(object_id, id_string, ID_STRING_SIZE));
-  ARROW_UNUSED(id_string);
+  std::string id_string = object_id.hex();
+  LOG_DEBUG("object table subscribe callback for OBJECT = %s", id_string.hex());
```
Comment (pcmoritz) [inline src/global_scheduler/global_scheduler.cc]: this .hex() is wrong

File: src/common/common.h [hunk 3]
```diff
@@ -113,118 +114,10 @@ extern "C" {
...
-
-/**
- * Compare two worker IDs.
- *
- * @param first_id The first worker ID to compare.
- * @param second_id The first worker ID to compare.
- * @return True if the worker IDs are the same and false otherwise.
- */
-bool WorkerID_equal(WorkerID first_id, WorkerID second_id);
+typedef ray::UniqueID WorkerID;
```
Comment (robertnishihara) [inline src/common/common.h]: Isn't this already defined in `id.h`?

File: src/common/common.h [hunk 4]
```diff
@@ -113,118 +114,10 @@ extern "C" {
...
-bool DBClientID_equal(DBClientID first_id, DBClientID second_id);
-
-/**
- * Compare a db client ID to the nil ID.
- *
- * @param id The db client ID to compare to nil.
- * @return True if the db client ID is equal to nil.
- */
-bool DBClientID_is_nil(ObjectID id);
+typedef ray::UniqueID DBClientID;
```
Comment (robertnishihara) [inline src/common/common.h]: Isn't this already defined in `id.h`?

File: src/common/logging.cc [hunk 5]
```diff
@@ -79,7 +79,7 @@ void RayLogger_log(RayLogger *logger,
 
     int status =
         redisAsyncCommand(context, NULL, NULL, formatted_message,
-                          (char *) db->client.id, sizeof(db->client.id));
+                          (char *) db->client.data(), sizeof(db->client));
```
Comment (robertnishihara) [inline src/common/logging.cc]: we should be using C++ style casts everywhere (at least for new code)

File: src/common/state/redis.cc [hunk 6]
```diff
@@ -672,7 +673,8 @@ void redis_object_table_lookup_callback(redisAsyncContext *c,
     for (size_t j = 0; j < reply->elements; ++j) {
       CHECK(reply->element[j]->type == REDIS_REPLY_STRING);
       DBClientID manager_id;
-      memcpy(manager_id.id, reply->element[j]->str, sizeof(manager_id.id));
+      memcpy(manager_id.mutable_data(), reply->element[j]->str,
```
Comment (robertnishihara) [inline src/common/state/redis.cc]: We have a decent number of `memcpy`'s in the codebase. Maybe this functionality should be part of the `ObjectID` class.

File: src/common/task.cc [hunk 7]
```diff
@@ -145,23 +145,23 @@ void free_task_builder(TaskBuilder *builder) {
 }
 
 bool TaskID_equal(TaskID first_id, TaskID second_id) {
```
Comment (robertnishihara) [inline src/common/task.cc]: The plan is for these `equal` and `is_nil` methods to go away (e.g., in a future PR) and just use `==` and `is_nil`, right?

File: src/global_scheduler/global_scheduler.cc [hunk 8]
```diff
@@ -339,11 +336,9 @@ void local_scheduler_table_handler(DBClientID client_id,
...
   GlobalSchedulerState *state = (GlobalSchedulerState *) user_context;
   ARROW_UNUSED(state);
-  char id_string[ID_STRING_SIZE];
-  LOG_DEBUG(
-      "Local scheduler heartbeat from db_client_id %s",
-      ObjectID_to_string((ObjectID) client_id, id_string, ID_STRING_SIZE));
-  ARROW_UNUSED(id_string);
+  std::string id_string = client_id.hex();
+  LOG_DEBUG("Local scheduler heartbeat from db_client_id %s",
+            id_string.c_str());
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: What's the plan for these logging macros down the road? E.g., Arrow uses something more compatible with C++ style logging, which would be nice here so we could avoid the `.c_str` and not have to worry so much about string formatting.

File: src/plasma/plasma_manager.h [hunk 9]
```diff
@@ -59,7 +59,7 @@ void PlasmaManagerState_free(PlasmaManagerState *state);
  * manager.
  */
 void process_transfer(event_loop *loop,
-                      ObjectID object_id,
+                      ray::ObjectID object_id,
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.h]: not super critical, but all of the `ray::ObjectID object_id` in this file should be `const ray::ObjectID &object_id`.

File: src/ray/id.cc [hunk 10]
```diff
@@ -22,16 +26,17 @@ UniqueID UniqueID::from_binary(const std::string &binary) {
...
   return result;
 }
 
-bool is_nil(const UniqueID &rhs) {
+bool UniqueID::is_nil() const {
   int i = kUniqueIDSize;
-  const uint8_t *data = rhs.data();
-  while (--i > 0 && data[i] == 255) {
+  const uint8_t *d = data();
+  while (--i > 0 && d[i] == 255) {
```
Comment (robertnishihara) [inline src/ray/id.cc]: I might be misreading this, but doesn't this fail to check `d[0]`? Also, we don't have any guarantees about the order in which the two boolean clauses are executed, right? But if they are executed out of order then we will have problems.

**Review by robertnishihara — COMMENTED** [review]
Great work here, this is way cleaner than before!

File: src/common/task.cc (hunk 7 above)
Comment (pcmoritz) [reply src/common/task.cc]: Yeah!

File: src/common/state/redis.cc (hunk 6 above)
Comment (pcmoritz) [reply src/common/state/redis.cc]: This code will go away as soon as the new GCS client code will be merged, and the rest should be modernized piece by piece.

File: src/global_scheduler/global_scheduler.cc (hunk 8 above)
Comment (pcmoritz) [reply src/global_scheduler/global_scheduler.cc]: Yeah, the code is already in ray/util/logging.h and the various components should be modernized as we go along.

File: src/ray/id.cc (hunk 10 above)
Comment (pcmoritz) [reply src/ray/id.cc]: Good catch!

File: src/plasma/plasma_manager.h (hunk 9 above)
Comment (pcmoritz) [reply src/plasma/plasma_manager.h]: Yeah, let's do this incrementally!

File: src/common/common.h (hunk 3 above)
Comment (pcmoritz) [reply src/common/common.h]: fixed

File: src/common/common.h (hunk 4 above)
Comment (pcmoritz) [reply src/common/common.h]: fixed

File: src/common/logging.cc (hunk 5 above)
Comment (pcmoritz) [reply src/common/logging.cc]: let's do this incrementally

File: src/common/common.h (hunk 3 above)
Comment (pcmoritz) [reply src/common/common.h]: actually, in id.h we have ray::WorkerID and ray::DBClientID; as most of the GCS code will get removed, I'm just going to keep this for now.

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**